File: Hard/212. Word Search II/WordSearchII.py
'''
LeetCode Hard: 79. Word Search
https://leetcode.com/problems/word-search/

Check WordSearchII.cpp for brief solution explanation.
'''
class TrieNode:
    def __init__(self):
        self.nxt = [None for i in range(26)]
        self.isEnd = False
        self.word = ""

class Solution:
    def dfs(self, r, c, board, cur, length):
        if board[r][c]=='*' or cur.nxt[ord(board[r][c])-ord('a')]==None or length>self.L:
            return
        else:
            cur = cur.nxt[ord(board[r][c])-ord('a')]
        if cur.isEnd:
            self.ans.append(cur.word)
            cur.isEnd = False
        t = board[r][c]
        board[r][c] = '*'
        for dr, dc in self.adj:
            newR, newC = r+dr, c+dc
            if 0<=newR<self.M and 0<=newC<self.N:
                self.dfs(newR, newC, board, cur, length+1)
        board[r][c] = t

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        self.M = len(board)
        self.N = len(board[0])
        self.L = 0
        self.ans = []
        self.adj = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        root = TrieNode()
        for i in range(len(words)):
            cur = root
            for j in words[i]:
                if cur.nxt[ord(j)-ord('a')]==None:
                    cur.nxt[ord(j)-ord('a')] = TrieNode()
                cur = cur.nxt[ord(j)-ord('a')]
            cur.isEnd = True
            cur.word = words[i]
            self.L = max(self.L, len(words[i]))
        for i in range(self.M):
            for j in range(self.N):
                if root.nxt[ord(board[i][j])-ord('a')]!=None:
                    self.dfs(i, j, board, root, 1)
        return self.ans

File: Hard/212. Word Search II/WordSearchII.java
/*
LeetCode Hard: 212. Word Search II
https://leetcode.com/problems/word-search-ii/

Check WordSearchII.cpp for brief solution explanation.
*/
class TrieNode{
    public TrieNode[]nxt;
    public boolean isEnd;
    public String word;
    public TrieNode(){
        nxt = new TrieNode[26];
        isEnd = false;
        word = "";
    }
}
class Solution {
    public int M, N;
    public int[][]adj = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public List<String>ans;
    public boolean bounds(int r, int c){
        return 0<=r&&r<M&&0<=c&&c<N;
    }
    public void dfs(int r, int c, char[][]board, TrieNode cur){
        if(board[r][c]=='*'||cur.nxt[board[r][c]-'a']==null){
            return;
        }else{
            cur = cur.nxt[board[r][c]-'a'];
        }
        if(cur.isEnd){
            ans.add(cur.word);
            cur.isEnd = false;
        }
        char t = board[r][c];
        board[r][c] = '*';
        for(int i = 0;i<4;i++){
            int newR = r+adj[i][0], newC = c+adj[i][1];
            if(bounds(newR, newC)){
                dfs(newR, newC, board, cur);
            }
        }
        board[r][c] = t;
    }
    public List<String> findWords(char[][] board, String[] words) {
        M = board.length;
        N = board[0].length;
        ans = new LinkedList<String>();
        TrieNode root = new TrieNode();
        for(int i = 0;i<words.length;i++){
            TrieNode cur = root;
            for(int j = 0;j<words[i].length();j++){
                if(cur.nxt[words[i].charAt(j)-'a']==null){
                    cur.nxt[words[i].charAt(j)-'a'] = new TrieNode();
                }
                cur = cur.nxt[words[i].charAt(j)-'a'];
            }
            cur.isEnd = true;
            cur.word = words[i];
        }
        for(int i = 0;i<M;i++){
            for(int j = 0;j<N;j++){
                if(root.nxt[board[i][j]-'a']!=null){
                    dfs(i, j, board, root);
                }
            }
        }
        return ans;
    }
}